app/src/test/java/mozilla/lockbox/presenter/AutofillLockedPresenterTest.kt

```diff
@@ -0,0 +1,168 @@
+
+    val subject = AutofillLockedPresenter(view, dispatcher, fingerprintStore, settingStore, lockedStore)
+
+    @Before
+    fun setUp() {
+        dispatcher.register.subscribe(dispatcherObserver)
+        subject.onViewReady()
+    }
+
+    @Test
+    fun `pushes autofill cancel action on unlockConfirmed(false)`() {
+        view._onUnlockConfirm.onNext(false)
+        dispatcherObserver.assertLastValue(AutofillAction.Cancel)
+    }
+
+    @Test
+    fun `pushes unlock action on unlockConfirmed(true)`() {
+        view._onUnlockConfirm.onNext(true)
+        dispatcherObserver.assertLastValue(DataStoreAction.Unlock)
+    }
+
+    @Test
+    fun `pushes autofill cancel action to finish with null on fingerprint auth canceled`() {
+        lockedStore.authenticationStub.onNext(FingerprintAuthAction.OnCancel)
+        dispatcherObserver.assertLastValue(AutofillAction.Cancel)
+    }
+
+    @Test
+    fun `on successful fingerprint authentication`() {
+        lockedStore.authenticationStub.onNext(FingerprintAuthAction.OnAuthentication(FingerprintAuthCallback.OnAuth))
+        dispatcherObserver.assertLastValue(DataStoreAction.Unlock)
+    }
+
+    @Test
+    fun `on unsuccessful fingerprint authentication when there is no other security`() {
+        fingerprintStore.keyguardDeviceSecureStub = false
+        lockedStore.authenticationStub.onNext(FingerprintAuthAction.OnAuthentication(FingerprintAuthCallback.OnError))
+        dispatcherObserver.assertLastValue(DataStoreAction.Unlock)
+    }
+
+    @Test
+    fun `on unsuccessful fingerprint authentication when there is other security`() {
+        fingerprintStore.keyguardDeviceSecureStub = true
+        lockedStore.authenticationStub.onNext(FingerprintAuthAction.OnAuthentication(FingerprintAuthCallback.OnError))
+        assertTrue(view.unlockFallbackCalled)
+    }
+
+    @Test
+    fun `when fingerprint auth is available and enabled`() {
+        fingerprintStore.fingerprintAuthAvailableStub = true
+        settingStore.unlockWithFingerprintStub.onNext(true)
+
+        Thread.sleep(600)
+        val fingerprintDialog = dispatcherObserver.values()[0] as RouteAction.DialogFragment.FingerprintDialog
+        assertEquals(R.string.fingerprint_dialog_title, fingerprintDialog.dialogTitle)
+    }
+
+    @Test
+    fun `when fingerprint auth is available but not enabled and there is other device security`() {
+        fingerprintStore.fingerprintAuthAvailableStub = true
+        fingerprintStore.keyguardDeviceSecureStub = true
+        settingStore.unlockWithFingerprintStub.onNext(false)
+        Thread.sleep(600)
+        assertTrue(view.unlockFallbackCalled)
+    }
+
+    @Test
+    fun `when fingerprint auth is not available or enabled and there is other device security`() {
+        fingerprintStore.fingerprintAuthAvailableStub = false
+        fingerprintStore.keyguardDeviceSecureStub = true
+        settingStore.unlockWithFingerprintStub.onNext(false)
+        Thread.sleep(600)
```

Comment:
🤢 Gosh do we have to?

Reply:
I knowwww, I was having issues making the `delay(0)` work and got frustrated >.>